### PR TITLE
Simplify paths in .hjson files used by dvsim.py

### PR DESCRIPTION
**Pull request by rswarbrick**
This patch series makes it so that we can vendor the contents of `hw/dv/data` and `hw/dv/tools` into Ibex properly, rather than using the current fragile "copy across and edit stuff" approach.

There are 3 patches in the series. Patches 1 and 2 are refactorings and code improvements in dvsim.py. Patch 3 defines the `dv_root` variable which makes it easier to vendor the hjson files.

After this patch series, together with the vendor tool changes in #2077, the Ibex project will be able to vendor in the contents of `hw/dv/{data,tools}` and only needs to patch 2 lines to get simulations working (the definition of `dv_root` in `dvsim.py` and the path to `common_project_cfg.hjson` in `common_sim_cfg.hjson`).

Edit: There used to be more patches in the series, but they have been dropped or done in other PRs.

### CODE REVIEW

**Comment by rswarbrick** [comment]
@weicaiyang Could you take a particular look at the "Tidy up config file loading in FlowCfg.py" patch? The command line override code was from you and I *think* it was just to support specifying `--tool` when running dvsim.py, but I'd be grateful if you'd check that I understood right.

**Comment by weicaiyang** [comment]
> > @weicaiyang Could you take a particular look at the "Tidy up config file loading in FlowCfg.py" patch? The command line override code was from you and I _think_ it was just to support specifying `--tool` when running dvsim.py, but I'd be grateful if you'd check that I understood right.
> 
> the tool override looks good to me. Thanks for the enhancement. @sriyerg please also take a look.
> 

Tried a few cases of using`--tool`, if we don't specify any tool, will see below error
`tool` can be either from hjson or command line override. command line has higher priority, but if it's not set at command line, need to get it from hjson file.
```
CRITICAL: [utils] Failed to parse "/usr/local/google/home/weicai/ot/uart/hw/dv/data//.hjson" possibly due to bad path or syntax error.
[Errno 2] No such file or directory: '/usr/local/google/home/weicai/ot/uart/hw/dv/data//.hjson'
```

File: util/dvsim/dvsim.py [hunk 1]
```diff
@@ -485,15 +485,20 @@ def main():
...
     # and other ASIC flow targets.
     if args.tool in ['ascentlint', 'veriblelint']:
-        cfg = LintCfg.LintCfg(args.cfg, proj_root, args)
+        cfg = LintCfg.LintCfg(args.cfg, proj_root, dv_root, args)
     elif args.tool == 'dc':
-        cfg = SynCfg.SynCfg(args.cfg, proj_root, args)
+        cfg = SynCfg.SynCfg(args.cfg, proj_root, dv_root, args)
     else:
-        cfg = SimCfg.SimCfg(args.cfg, proj_root, args)
+        cfg = SimCfg.SimCfg(args.cfg, proj_root, dv_root, args)
```
[the remark this answers is not in the document]
Comment (rswarbrick) [reply util/dvsim/dvsim.py]: Good point. I'll move it.

**Comment by rswarbrick** [comment]
(force-push to add missing sign-off line)

**Comment by rswarbrick** [comment]
Oh, sorry, @weicaiyang, but I forgot to address your `use_cfgs` comment. As far as I know, I haven't changed dvsim's behaviour with `use_cfgs`: the code has been re-arranged slightly, but it shouldn't have changed meaning.

File: util/dvsim/FlowCfg.py [hunk 2]
```diff
@@ -99,11 +99,11 @@ def __post_init__(self):
                 log.error("Parse error!\n%s", self.cfgs)
                 sys.exit(1)
 
-    @staticmethod
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:102]: Thanks for this change. Smushing different unrelated updates into the same PR makes the review hard. Would you please split this out to a separate PR?

Comment (rswarbrick) [reply util/dvsim/FlowCfg.py:102]: Oh, good point. The original version of this PR used this (because it passed a `dv_root` parameter around), but it's not needed since @weicaiyang's suggestion. I'll split this out now.

**Comment by rswarbrick** [comment]
Force-push rebases and then removes the first patch, which is no longer needed (and has been split into a separate PR).

**Comment by rswarbrick** [comment]
Hi @sriyerg. I agree that moving everything into the dvsim directory and setting a `dvsim_root` variable would work. But I think it's a much bigger change, and the stuff in this patch series is enough for what we need right now. Can we punt on the big restructure until the future?

**Comment by rswarbrick** [comment]
Hi @sriyerg. Could you have another look at this PR and let me know what remaining changes you're requesting? I don't think I can tell from reading through the comments above and I feel like I've got quite a pile of dvsim.py PRs piling up. Thanks!

**Comment by rswarbrick** [comment]
Hi again, @sriyerg. Any chance you could have another look at this PR before I rebase it?

File: util/dvsim/FlowCfg.py [hunk 3]
```diff
@@ -18,6 +18,10 @@
 # A set of fields that can be overridden on the command line.
 _CMDLINE_FIELDS = {'tool'}
 
+# A set of fields to resolve as paths, relative to the hjson file that sets
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:21]: Hiding the detail of which hjson keys are 'relative' paths in the python code does not seems like a good approach to me. I would prefer to keep the existing approach (use {proj_root} or {dvsim_tool_root} as anchors to specify full paths to files).

File: util/dvsim/SimCfg.py [hunk 4]
```diff
@@ -120,6 +120,15 @@ def __init__(self, flow_cfg_file, proj_root, args):
         # Process overrides before substituting the wildcards.
         self._process_overrides()
 
+        # Make sure that we have a tool by this point. We support calling
+        # simulations with no tool specified on the command line but in that
+        # case there needs to be a default specified in the config file, which
+        # should have been set by this point.
+        if not self.tool:
```
Comment (sriyerg) [inline util/dvsim/SimCfg.py:132]: This check is valid only for an individual cfg, not for a master cfg. Please move this into the `if` clause on line 149.

File: util/dvsim/FlowCfg.py (hunk 3 above)
Comment (rswarbrick) [reply util/dvsim/FlowCfg.py]: Ok, I've thrown out the "Read flow_makefile relative..." patch and changed the import in `common_sim_cfg.hjson` to use `dv_root`.

File: util/dvsim/SimCfg.py (hunk 4 above)
Comment (rswarbrick) [reply util/dvsim/SimCfg.py]: Done.

**Comment by rswarbrick** [comment]
Everything is rebased onto master again. This drops the --tool handling code, which was done in e83b55e (in a different PR).

**Comment by rswarbrick** [comment]
Force-push rebases onto master again. This patch touches lots of files (in trivial ways), so needs frequent rebasing. @sriyerg please could you have another look at this so we can get it merged? I'm also keen to tidy up how we vendor stuff into Ibex, which depends on this work.

File: util/dvsim/SimCfg.py [hunk 5]
```diff
@@ -112,6 +112,12 @@ def __init__(self, flow_cfg_file, proj_root, args):
         self.dry_run = args.dry_run
         self.map_full_testplan = args.map_full_testplan
 
+        # Choose a value for dv_root. This is currently always at
+        # <proj_root>/hw/dv but at least listing it here rather than in the
+        # various hjson files means that other projects vendoring us in can
+        # patch the path simply.
+        self.dv_root = os.path.join(proj_root, 'hw', 'dv')
```
Comment (sriyerg) [inline util/dvsim/SimCfg.py:119]: The tool should make no assumptions on what paths exist. Please remove these lines. Instead add this to `hw/data/common_project_cfg.hjson`. I would imagine the Ibex repo will provide its own 'project' specific hjson that sets the `dv_root` correctly as needed.

Comment (rswarbrick) [reply util/dvsim/SimCfg.py:119]: The problem is that `common_project_cfg.hjson` is included by `common_sim_cfg.hjson`, which is explicitly named in each IP's hjson file and is found relative to `dv_root`. It feels a little ugly to replace `{proj_root}/hw/dv` with `{dv_root}`, except if it happens to be the inclusion of `common_sim_cfg.hjson`.

Is that what you would like to do? Or can you think of another trick?

**Comment by rswarbrick** [comment]
Hi @sriyerg! I know that we don't have much bandwidth for reviewing dvsim code at the moment, but this series of patches makes it possible to vendor better in Ibex and they have now been stalled for 3 weeks. I've just rebased onto master: could you take a look and see whether there are any other changes needed?

**Comment by sjgitty** [comment]
Just pinging @sriyerg here again as this has been pending for a week.

**Comment by rswarbrick** [comment]
Fortnightly re-ping for @sriyerg. You last commented on this on June 10th.

**Comment by rswarbrick** [comment]
Force-push rebases onto master (it was conflicting with the recent master -> primary renaming).

**Comment by rswarbrick** [comment]
It's been (yet) another fortnight. @sriyerg: please would you review these changes or at least respond to this PR?

At the moment, it seems that no-one else is allowed to review dvsim patches. Since you are ignoring this PR, it's impossible for it to land. I realise that this isn't a priority for you, but it's been almost 2 months since you last replied and I'd really like to merge this so that I can do the corresponding tidy-ups in Ibex.

**Comment by sriyerg** [comment]
Apologies it has taken this long. Reviewing now.

File: util/dvsim/FlowCfg.py [hunk 6]
```diff
@@ -153,94 +156,103 @@ def check_if_primary_cfg(self, hjson_dict):
...
+            setattr(self, key, dict_val)
+            return
+
+        # Neither value looks like a default. Raise an error.
+        log.error("When loading {!r}, key {!r} is given a value of "
+                  "{!r}{}, but the key is already set to {!r}."
+                  .format(path, key, dict_val, override_msg, self_val))
+        sys.exit(1)
+
+    def resolve_hjson_raw(self, path, hjson_dict):
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:233]: The newly added `path` argument (to this and to `_set_attribute` methods) serves no other purpose than to make the error messages clear. Is that correct?

File: util/dvsim/FlowCfg.py [hunk 7]
```diff
@@ -15,6 +15,9 @@
 from Deploy import Deploy
 from utils import VERBOSE, md_results_to_html, parse_hjson, subst_wildcards
 
+# A set of fields that can be overridden on the command line.
+_CMDLINE_FIELDS = {'tool'}
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:19]: All command line args should take precedence over what is set in the hjson files. This IMO should not be needed.

**Comment by sriyerg** [comment]
I tried running the following, but ran into the following error:
```console
./util/dvsim/dvsim.py hw/top_earlgrey/dv/top_earlgrey_sim_cfgs.hjson -i sanity -n 

CRITICAL: [utils] Failed to parse "/usr/local/google/home/sriyer/opentitan/repo0/opentitan/hw/dv/data//.hjson" possibly due to bad path or syntax error.
[Errno 2] No such file or directory: '/usr/local/google/home/sriyer/opentitan/repo0/opentitan/hw/dv/data//.hjson'
```

This can be fixed by removing the `default` for the `--tool` switch in `dvsim.py`.

File: util/dvsim/FlowCfg.py (hunk 7 above)
Comment (sriyerg) [reply util/dvsim/FlowCfg.py:19]: On second thoughts, I am ok keeping this in for now. Will figure out later what needs to be done here.

**Comment by sriyerg** [comment]
I 

> The following command fails:
> 
> ```
> ./util/dvsim/dvsim.py hw/top_earlgrey/dv/top_earlgrey_sim_cfgs.hjson -i sanity -n 
> 
> CRITICAL: [utils] Failed to parse "/usr/local/google/home/sriyer/opentitan/repo0/opentitan/hw/dv/data//.hjson" possibly due to bad path or syntax error.
> [Errno 2] No such file or directory: '/usr/local/google/home/sriyer/opentitan/repo0/opentitan/hw/dv/data//.hjson'
> ```
> 
> This can be fixed by removing the `default` for the `--tool` switch in `dvsim.py`.

This is now fixed with #3041.

**Comment by sriyerg** [comment]
LGTM, save for a question and rebase+push.

File: util/dvsim/FlowCfg.py (hunk 6 above)
Comment (rswarbrick) [reply util/dvsim/FlowCfg.py:233]: Yes, that's right. Since we have configuration files that can include each other (and the hjson library doesn't annotate its parsed data with line information) I think the very least we can do is tell the poor user which of their files has an error :-)

File: hw/data/common_project_cfg.hjson [hunk 8]
```diff
@@ -6,6 +6,9 @@
   doc_server:       docs.opentitan.org
   results_server:   reports.opentitan.org
 
+  // Where to find DV code
+  dv_root:            "{proj_root}/hw/dv"
```
[the remark this answers is not in the document]
Comment (rswarbrick) [reply hw/data/common_project_cfg.hjson]: Yep, I'll do that. I (wrongly) thought that `common_sim_cfg` was included by `common_project_cfg`, which would mean we needed `dv_root` to figure out where the file was! But that's wrong: it's included by each IP block's testsuite.
